File: programmers/60059.java
import java.util.*;

class Solution {

    boolean isUnlocked(int[][] expLock, int N, int M) {
        for (int i = M - 1; i < N + M - 1; i++) {
            for (int j = M - 1; j < N + M - 1; j++) {
                if (expLock[i][j] > 1) return false;
                if (expLock[i][j] == 0) return false;
            }
        }

        return true;
    }

    int[][] rotateArray(int[][] arr, int d) {
        int size = arr.length;
        int[][] rotated = new int[size][size];

        if (d % 4 == 1) {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < size; j++) {
                    rotated[j][size - 1 - i] = arr[i][j];
                }
            }
        }
        else if (d % 4 == 2) {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < size; j++) {
                    rotated[size - 1 - i][size - 1 - j] = arr[i][j];
                }
            }
        }
        else if (d % 4 == 3) {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < size; j++) {
                    rotated[size - 1 - j][i] = arr[i][j];
                }
            }
        }
        else {
            for (int i = 0; i < size; i++) {
                for (int j = 0; j < size; j++) {
                    rotated[i][j] = arr[i][j];
                }
            }
        }


        return rotated;
    }

    int[][] match(int startX, int startY, int M, int[][] expLock, int[][] key) {
        int lockLen = expLock.length;
        int[][] tmpExpLock = new int[lockLen][lockLen];

        for (int i = 0; i < lockLen; i++) {
            for (int j = 0; j < lockLen; j++) {
                tmpExpLock[i][j] = expLock[i][j];
            }
        }

        for (int i = startX; i < startX + M; i++) {
            for (int j = startY; j < startY + M; j++) {
                tmpExpLock[i][j] += key[i - startX][j - startY];
            }
        }

        return tmpExpLock;
    }

    boolean search(int startX, int startY, int N, int M, int[][] expLock, int[][] key) {

        int[][] matchedExpLock = match(startX, startY, M, expLock, key);
        if (isUnlocked(matchedExpLock, N, M)) return true;

        for (int i = 1; i < 4; i++) {
            int[][] rotatedKey = rotateArray(key, i);
            matchedExpLock = match(startX, startY, M, expLock, rotatedKey);

            if (isUnlocked(matchedExpLock, N, M)) return true;
        }

        return false;
    }

    public boolean solution(int[][] key, int[][] lock) {
        boolean answer = true;

        int M = key.length;
        int N = lock.length;
        int[][] expLock = new int[N + (M - 1) * 2][N + (M - 1) * 2];
        for (int i = 0; i < expLock.length; i++) {
            for (int j = 0; j < expLock.length; j++) {
                expLock[i][j] = -1;
            }
        }

        for (int i = M - 1; i < N + M - 1; i++) {
            for (int j = M - 1; j < N + M - 1; j++) {
                expLock[i][j] = lock[i - (M - 1)][j - (M - 1)];
            }
        }

        for (int i = 0; i < N + M - 1; i++) {
            for (int j = 0; j < N + M - 1; j++) {
                if (search(i, j, N, M, expLock, key)) return true;
            }
        }

        return false;
    }
}

File: programmers/60059.swift
import Foundation

func rotateMap(_ map: [[Int]]) -> [[Int]] {
    let n = map.count
    var rotated = map
    
    for i in 0..<n {
        for j in 0..<n {
            rotated[i][j] = map[n-j-1][i]
        }
    }
    return rotated
}

func isValid(_ map: [[Int]]) -> Bool {
    let m = map.count/3
    
    for r in m..<m*2 {
        for c in m..<m*2 {
            if map[r][c] == 0 || map[r][c] == 2 {
                return false
            }
        }
    }
    return true
}

func solution(_ key:[[Int]], _ lock:[[Int]]) -> Bool {
    let n = key.count // 열쇠의 크기
    let m = lock.count // 자물쇠의 크기
    
    // 3배로 확장한 자물쇠 배열 만들기
    var extended = [[Int]](repeating: [Int](repeating: 0, count: m*3), count: m*3)
    
    for i in 0..<m {
        for j in 0..<m {
            extended[i+m][j+m] = lock[i][j]
        }
    }
    
    for i in 0..<m*2 {
        for j in 0..<m*2 {
            var rotated = key
            for k in 0..<4 {
                // 열쇠 배열을 돌린다
                rotated = rotateMap(rotated)
                
                // 열쇠 배열을 끼운다
                var copied = extended
                for r in 0..<n {
                    for c in 0..<n {
                        copied[i+r][j+c] += rotated[r][c]
                    }
                }
                // 잘 맞는지 확인
                if isValid(copied) { return true }
            }
        }
    }
    
    
    return false
}
